src/utils/formatting.py

```
@@ -0,0 +1,258 @@
+    Args:
+        conclusion: The main conclusion
+        premises: List of supporting premises
+
+    Returns:
+        Formatted observation string
+    """
+    premises_text = format_premises_for_display(premises)
+    return f"{conclusion}{premises_text}"
+
+
+def extract_observation_content(observation: str | dict[str, Any] | Any) -> str:
```

Ruff (0.12.2)

Dynamically typed expressions (typing.Any) are disallowed in `observation`

(ANN401)